### ISSUE TITLE
Need downgrade tests for 1.8 release

### ISSUE THREAD

**Issue by gmarek** [kubernetes/kubernetes#52593]
From what I learned on slack we're testing upgrades only - we check if we upgrade 1.7 cluster to 1.8 cluster things keep working. On the other hand we're NOT testing if downgrading 1.8 cluster to 1.7 works at all. Downgrades are extremely important for everyone running kubernetes in production - if it turns out that 1.8 has a bug that breaks some of user's workloads.

Standard mitigation policy in such case is to downgrade cluster to 1.7 and start working on a fix. In the absence of downgrade possibility, he/she will need to wait, possibly days, for hot-fix which is risky by itself.

We should consider making downgrades work a prerequisite for releasing 1.8. I'm aware that we haven't done downgrade testing since around 1.5, but the fact that we somehow managed that until now doesn't mean we should continue to ignore this problem.

@kubernetes/sig-release-bugs @thockin @bgrant0607 @smarterclayton @spiffxp

**Comment by smarterclayton** [comment]
I agree, even a manual test would be better than nothing.  The potential impact of *not* knowing what would happen puts the project at risk.

**Comment by eparis** [comment]
Upgrades can do storage layer re-writes, which would be downgrade incompatible. Right? Isn't 'downgrades working' basically impossible over the long run?

**Comment by smarterclayton** [comment]
Technically every API field has to be forward and backwards compatible, so new data would be lost, but the system should function without that data.  So normal objects that don't have their storage version changed would be unaffected.  However, if the upgrade has changed what the preferred storage version is (such as deployments in apps/v1beta2), then yes, any "new" versions would be unreadable after a downgrade.

**Comment by gmarek** [comment]
This means we should consider having a tool that "moves" the storage of a given type to different (compatible) one.

**Comment by smarterclayton** [comment]
We do have a variant of it, but need to get more formal about it.

**Comment by jdumars** [comment]
This testing is going to be like putting a pair of pants full of loose change in the dryer.  We're going to find a lot of work shaking out of this.  The worst thing we can do is be blind to what that work may be, because a no-revert strategy for cluster operators is not a good UX and will inevitably lead to a decline in adoption.

**Comment by thockin** [comment]
I like the analogy, @jdumars.

I wrote https://github.com/kubernetes/kubernetes.github.io/pull/5338 specifically to address the issues around default storage version, which lead to https://github.com/kubernetes/kubernetes/issues/52185

**Comment by bgrant0607** [comment]
Related: #52185, #4855, #3562, #46073, #23233

**Comment by bgrant0607** [comment]
One known issue: Any resources created using an API that first appears (enabled) in a release will be orphaned in etcd when rolling back, which then become time bombs for when the cluster is rolled forward again, kind of similar to the annotations problem described in #30819.

**Comment by dims** [comment]
@gmarek @bgrant0607 @thockin - Other than https://github.com/kubernetes/test-infra/pull/4608 do we need to do anything else in 1.8? it's a bit late already

**Comment by krzyzacy** [comment]
[master-1.7 downgrade](https://k8s-testgrid.appspot.com/master-upgrade#gke-gci-master-gci-1.7-downgrade-cluster) is green (Thanks @krousey! ), will set up 1.8-1.7 downgrade job later.
Also need to put up a suite test slow/serial jobs.

**Comment by gmarek** [comment]
Huge thanks @krzyzacy!

**Comment by jurgenweber** [comment]
I have had to downgrade from 1.8 to 1.7..... but I am now getting this error in relation to the batchjobs/cronjobs API:

``` kubectl version
Client Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.10", GitCommit:"bebdeb749f1fa3da9e1312c4b08e439c404b3136", GitTreeState:"clean", BuildDate:"2017-11-03T17:13:44Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.10", GitCommit:"bebdeb749f1fa3da9e1312c4b08e439c404b3136", GitTreeState:"clean", BuildDate:"2017-11-03T16:31:49Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
 $ kubectl get cronjobs -n devops
Error from server: no kind "CronJob" is registered for version "batch/v1beta1"
```

any ideas on how to get hte cronjob api back?

**Comment by jurgenweber** [comment]
Ok, I found a solution. I found the entries for the cronjobs in etcd-server had 

```
PrevNode.Value: {"kind":"CronJob","apiVersion":"batch/v1beta1","metadata" .....
```
I deleted them from etcd and recreated them and they are back.
